scripts/cwl_workflows/lnd-elm/ncremap_lnd.cwl

```diff
@@ -15,19 +15,12 @@ arguments:
   - ncremap
   - -P
   - "elm"
```

Comment:
Quotes on elm is not needed. I think we should add other flags, similar to eam:
  - -7
  - --dfl_lvl=1
  - --no_cll_msr
  - --no_stdin

Reply:
I can update and test.  As far as "n2n" and "elm", I note that our v1 work (in datasm, as written by Sterling) called "n2n", and v2 called "elm".  BOTH were written with the "-s -g -sgs_frc" method, and having converted both to "-P elm, -m mapfile", the only remaining difference is that n2n searches for files with "clm2", while elm uses "elm".  If we can pass "search_patt: clm2" or "search_patt: elm" as a parameter to the workflow - both become identical.

Reply:
Well, this is unfortunate:  I added below "ncremap"

  - ncremap
  - -7
  - --dfl_lvl=1
  - --no_cll_msr
  - --no_stdin

But got never-before-seen errors:

ERROR Tool definition failed initialization:
Tool definition file:///p/user_pub/e3sm/staging/resource/cmor/cwl_workflows/lnd-elm/ncremap_lnd.cwl failed validation:
../../../../../../staging/resource/cmor/cwl_workflows/lnd-elm/ncremap_lnd.cwl:2:1:     Object
                                                                                      `../../../../../../staging/resource/cmor/cwl_workflows/lnd-elm/ncremap_lnd.cwl`
                                                                                      is not valid because
                                                                                        tried
                                                                                        `CommandLineTool`
                                                                                        but
../../../../../../staging/resource/cmor/cwl_workflows/lnd-elm/ncremap_lnd.cwl:8:1:         the `arguments`
                                                                                          field is not valid
                                                                                          because
                                                                                            tried array
                                                                                            of <string or
                                                                                            Expression or
                                                                                            CommandLineBinding>
                                                                                            but
../../../../../../staging/resource/cmor/cwl_workflows/lnd-elm/ncremap_lnd.cwl:16:5:           item is
                                                                                              invalid because
                                                                                                - tried
                                                                                                string but
                                                                                                  the
                                                                                                  value is not string
                                                                                                - tried
                                                                                                Expression but
                                                                                                  value
                                                                                                  is a int but
                                                                                                  expected a string

Reply:
I suspect it could be a syntax error? maybe following this example. https://github.com/E3SM-Project/e3sm_to_cmip/blob/61d7fdf4d17f8ce814b9bc786d1c1c089221968a/scripts/cwl_workflows/atm-unified-eam/hrzremap_posin.cwl

Reply:
Regarding
  - ncclimo
  - '-7'
  - --dfl_lvl=1
  - --no_cll_msr
  - --no_stdin
  - -P
  - "eam"
  
  As you suggested, the quotes on '"eam"' are not needed (works fine either way.)
  Single-quoting (maybe double as well) the '- 7' worked.  But including  the "--no_cll_msr" led to the error from ncap2
  
  ncap2: ERROR out(): Could not read var area
  and 
  ncremap: ERROR Failed post-processing to convert output area from [sr] to [km2]. Debug this:
ncap2 -O -s 'area*=6.37122e6^2/1.0e6;area@long_name="Gridcell area";area@units="km^2"' "/p/user_pub/e3sm/bartoletti1/Pub_Work/2_PostProcess/altprocess/ManualTest/ncremap/tmp_u2v6xgd/v2.LR.abrupt-4xCO2_0101.e
lm.h0.0001-01.nc" "/p/user_pub/e3sm/bartoletti1/Pub_Work/2_PostProcess/altprocess/ManualTest/ncremap/tmp_u2v6xgd/v2.LR.abrupt-4xCO2_0101.elm.h0.0001-01.nc"

I'll skip that one, and see if it complains about "--no_stdin"

Reply:
And of course it does.  Changing

```
  - -P
  - elm
  - -m
  - $(inputs.remapfile)
  - -v
  - $(inputs.var_list.join(','))

inputs:
  lnd_files:
    type: File
  account:
    type: string
  partition:
    type: string
  timeout:
    type: string
  remapfile:
    type: string
  var_list:
    type: string[]

stdin: $(inputs.lnd_files.path)
```

to
```

  - -P
  - elm
  - -m
  - $(inputs.remapfile)
  - -v
  - $(inputs.var_list.join(','))
  - $(inputs.lnd_files.path)

inputs:
  lnd_files:
    type: File
  account:
    type: string
  partition:
    type: string
  timeout:
    type: string
  remapfile:
    type: string
  var_list:
    type: string[]
```

Produced the obscure error

    -P \
    elm \
    -m \
    /p/user_pub/e3sm/staging/resource/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc \
    -v \
    TSOI \
    /p/user_pub/e3sm/bartoletti1/Pub_Work/2_PostProcess/altprocess/ManualTest/ncremap/tmpqiwbh3fy/stgcce0aaff-ebe8-4532-96a1-1e2c56555781/lnd_files.txt
Input #00: /p/user_pub/e3sm/bartoletti1/Pub_Work/2_PostProcess/altprocess/ManualTest/ncremap/tmpqiwbh3fy/stgcce0aaff-ebe8-4532-96a1-1e2c56555781/lnd_files.txt
Map/Wgt  : /p/user_pub/e3sm/staging/resource/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
nco_err_exit(): ERROR Short NCO-generated message (usually name of function that triggered error): nco__open()
nco_err_exit(): ERROR Error code is -51. Translation into English with nc_strerror(-51) is "NetCDF: Unknown file format"
ERROR: nco__open() unable to open file "/p/user_pub/e3sm/bartoletti1/Pub_Work/2_PostProcess/altprocess/ManualTest/ncremap/tmpqiwbh3fy/stgcce0aaff-ebe8-4532-96a1-1e2c56555781/lnd_files.txt"
ERROR NC_ENOTNC Not a netCDF file

And that file disappears as soon as it is created:

(e2c_dev2) ll tmpqiwbh3fy/stgcce0aaff-ebe8-4532-96a1-1e2c56555781/lnd_files.txt
ls: cannot access tmpqiwbh3fy/stgcce0aaff-ebe8-4532-96a1-1e2c56555781/lnd_files.txt: No such file or directory

BTW, the example you gave is for ncclimo, not ncremap, if that matters.

Reply:
Also, curiously, the ncclimo variables input (example) says

 - $(inputs.variables.join(" "))        [space-separated]
 
 whereas the variables input supplied to ncremap are comma-separated.
 
   - $(inputs.var_list.join(','))

Reply:
Tony, are you saying that neither `--no_cll_msr` nor `--no_stdin` works?

Reply:
Jill, the `--no_cll_msr` caused a `ncap2: ERROR out(): Could not read var area`.

The `--no_stdin` might work, but it will force you to revamp the workflow in ways in which I am unfamiliar.  You certainly cannot just add it to the command line.  The existing workflow "lnd.cwl" has subordinate workflows "discover_lnd_files.cwl", whose output is input to "ncremap_lnd.cwl", whose output is input to ""timeseries.cwl"", whose output is input to "cmor.cwl".  Each of these has a peculiar way of specifying input and output.

The existing "ncremap_lnd.cwl" has an ncremap command line that terminates with the "var list".  But when it is run, the logfile shows it ends with

```
-v \
    TSOI < /p/user_pub/e3sm/bartoletti1/Pub_Work/2_PostProcess/altprocess/ManualTest/ncremap/tmp0fqntji7/stg3c63d5e6-2b79-45d6-9917-e4c0542af9ee/lnd_files.txt
```
and this works fine (don't ask me how.  I assume "lnd_files.txt" is a listing of the files.)

The script has the line:

`stdin: $(inputs.lnd_files.path)`
    
So I assumed I could add the "--no_stdin" and just add "$(inputs.lnd_files.path)" to the command line.  But that does not work.  If CWL supported print statements I could inspect some of these terms, but it does not.  I suppose I could create a "print.cwl" workflow, and call it as a separate step, giving it something I want to see printed - ridiculous...

Reply:
Thanks for clarifying. I was not aware that the `eam` workflow calls `ncclimo`, while the `elm` workflow calls `ncremp`. I thought both can have a same ncclimo/ncremp call. I think you can push the code change (leaving out --no_cll_msr and --no_stdin. And I will tag Charlie for a final pass. Thank you for experimenting.